libraries/rush-lib/src/logic/policy/EnvironmentPolicy.ts

```diff
@@ -0,0 +1,45 @@
+// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
+// See LICENSE in the project root for license information.
+
+import { AlreadyReportedError, Async, FileSystem } from '@rushstack/node-core-library';
+
+import type { RushConfiguration } from '../../api/RushConfiguration';
+import type { IPolicyValidatorOptions } from './PolicyValidator';
+import { RushConstants } from '../RushConstants';
+
+/**
+ * Ensures the environment where the Rush repo exists is valid
+ */
+export async function validateAsync(
+  rushConfiguration: RushConfiguration,
+  options: IPolicyValidatorOptions
+): Promise<void> {
+  if (rushConfiguration.experimentsConfiguration.configuration.forbidPhantomResolvableNodeModuleFolders) {
+    const pathParts: string[] = rushConfiguration.rushJsonFolder.split(/[\/\\]/);
+    const existingNodeModulesPaths: string[] = [];
+    await Async.forEachAsync(
+      pathParts,
+      async (pathPart: string, index: number) => {
+        const potentialNodeModulesPath: string = `${pathParts.slice(0, index + 1).join('/')}/node_modules`;
+        const pathExists: boolean = await FileSystem.existsAsync(potentialNodeModulesPath);
+        if (pathExists) {
+          existingNodeModulesPaths.push(potentialNodeModulesPath);
+        }
+      },
+      { concurrency: 5 }
+    );
```

Comment:
While this is probably cheap enough on account of occurring exactly once, I still would prefer looping on `path.dirname` or such instead of split/join. Not going to block on this, though.

Reply:
Is it even faster? To test `/x/y/z/` the OS has to traverse `/x/y/z` already. And then if an error occurs (e.g. `x` cannot be read) then we'd get 3 errors in parallel instead of just one.  And now you probably need to sort `existingNodeModulesPaths` because won't the promises resolve in unpredictable order?  For a significant performance gain, this complexity would be justified. But maybe it's just overengineering?

Reply:
So, running this 1000 times with parrelism took 77ms and without parallelism it took 108ms. I think that's worth it.

I'm less concerned with the perf or bubbling of filesystem read error cases.